### ISSUE TITLE
time of flight / Delayed differential equations

### ISSUE THREAD

**Issue by cdaunt** [cdaunt/circulax#2]
Currently components do not themselves account for a group delay.

This is usually handled by tracking the history (or buffers). Support for this is planned to be added to diffrax.

https://github.com/patrick-kidger/diffrax/pull/169

Group delays can be added to components once that feature has been added to diffrax

**Comment by cdaunt** [comment]
## Proposal: constant time-delay support without waiting on diffrax's DDE support

diffrax's DDE work (#169 upstream) is stalled on general **state-dependent** delays with discontinuity root-finding, and only handles `ConstantStepSize` so far. Circulax doesn't need that generality — a waveguide/transmission-line group delay is a fixed, known-in-advance value (`length / v_group`), not solution-dependent. Also, circulax no longer calls `diffrax.diffeqsolve` directly — `circuit_diffeq.py` is already a from-scratch integrator loop, so this can be added without any diffrax dependency.

**Scope:** fixed per-instance delay τ (can be a differentiable function of component params), constant step size for v1. Not solving state-dependent delays or discontinuity tracking.

**Mechanism:** `CircuitState` already carries preallocated output buffers filled via `.at[idx].set()` on each accepted step, registered with `eqxi.while_loop`'s `buffers=` so checkpointed AD handles them for free (see `SaveState.ts/.ys`). Add one more such buffer:

- `hist_t: Array[max_steps+1]`, `hist_y: Array[max_steps+1, sys_size]`, written on every accepted step.
- Delayed read at `t1 - τ`: `jnp.interp(t1 - τ, hist_t, col)` per state column — ascending `hist_t` (inf-padded tail) makes this a correct, cheap, differentiable lookup with sensible boundary behavior (τ → before t0 clamps to the initial condition).

**Component API:** follows the existing `init`/`.setup` injection pattern in `base_component.py` — a new reserved arg `hist`, injected with the delayed port-voltage tuple at `t - τ(params)`.

**Critical detail for per-instance-varying τ (e.g. waveguides of different lengths in one group):** the delayed read must vmap the *interpolation itself* over the instance axis, not compute one global delayed vector and slice it (that only works for uniform τ):

```python
def _delayed_read(tau_i, var_idx_i):
    tq = t1 - tau_i
    cols = hist_y[:, var_idx_i]
    return jax.vmap(lambda c: jnp.interp(tq, hist_t, c), in_axes=1)(cols)

hist_locs = jax.vmap(_delayed_read, in_axes=(0, 0))(tau, group.var_indices)
```

This composes with the existing per-group vmap (`assembly.py`) using the same `in_axes` shape. Cost is `O(N_instances × n_vars_per_instance)` interp calls, one fused XLA op — fine at typical circuit sizes. Gradients flow both w.r.t. τ (via `jnp.interp`'s query-point derivative) and w.r.t. history values (gather vjp).

**Constraints:** τ should be ≳ a few dt (guard via `eqxi.error_if`); v1 requires `ConstantStepSize`; buffer memory is `O(max_steps × sys_size)` — fine for typical `max_steps` but worth bounding later with a ring buffer sized off a `max_delay` hint instead of full trajectory.

**Delivery plan:**
1. `hist_t`/`hist_y` buffer in `CircuitState`/`_circuit_loop`, gated behind a `has_delay` flag on component groups (parallel to `is_fdomain`) so undelayed circuits pay nothing.
2. `hist` injection in `base_component.py` + one worked component (delay-line / `OpticalWaveguide` variant).
3. Assembly wiring (vmapped per-instance read above) + τ guard.
4. Tests: step response through a delay line vs. analytically shifted signal; gradient check w.r.t. `length_um`.
5. Follow-up (not blocking): adaptive step size + bounded ring buffer.

**Comment by cdaunt** [comment]
Phase 1 implemented and pushed: https://github.com/gdsfactory/circulax/tree/worktree-feat-time-delay (commit https://github.com/gdsfactory/circulax/commit/e76e087db28cb5d9ab2acbc61ec7f87e2dc04363).

Covers: `hist`/`.delay` component API, the accepted-step history buffer + per-instance vmapped `jnp.interp` read (handles varying `tau` across instances of one component type), `OpticalDelayLine`, and the `tau >= dt` / ConstantStepSize-only guards. 221 tests passing (216 existing + 5 new in `tests/test_delay.py`), including a gradient-vs-finite-difference check through the checkpointed adjoint.

No PR yet -- happy to open one when ready.

**Comment by cdaunt** [comment]
Follow-up: adaptive step size + delay is not supported in the current implementation (`setup_transient` raises `ValueError` if a non-`ConstantStepSize` controller is used with delayed components).

Why it's scoped out for v1, not just an oversight:
- The delay-history read (`jnp.interp` over accepted `(t, y)` samples) doesn't itself require uniform step size -- it should work fine against an irregular grid -- but this hasn't been exercised/tested against adaptive stepping.
- More importantly, the `tau >= dt` guard is checked against whatever `dt` the controller picks for the *current* step. With adaptive control, `dt` can shrink for reasons unrelated to delay (Newton convergence, stiffness elsewhere in the circuit), which could trip that guard unpredictably mid-run for a circuit that was fine at the start -- much harder for a user to reason about than a fixed `dt0` they chose themselves.
- Upstream diffrax's own general DDE support (referenced in this issue) has the same scoping: only `ConstantStepSize` handled so far, because coordinating step-size adaptation with delay-crossing/discontinuity handling is genuinely complex.

A real follow-up would need: removing the blanket rejection, deciding how the controller should behave when a proposed step would push `dt` below the smallest active `tau` (likely clamp the step similarly to how `t1` is already clamped in the loop), and validating `jnp.interp`'s accuracy against an irregularly-spaced history buffer.

**Comment by cdaunt** [comment]
Adaptive step size + delay follow-up implemented and pushed to https://github.com/gdsfactory/circulax/tree/worktree-feat-time-delay (commit https://github.com/gdsfactory/circulax/commit/5d37c8ffb7024fb93cd17c12642c5d943dce4e32).

Removed the blanket `ValueError` that rejected non-`ConstantStepSize` controllers on delayed circuits. Instead, `setup_transient()` now computes the smallest active `tau` across all delayed groups (`assembly.min_active_tau`) and, only when the controller is adaptive, proactively clamps the proposed step (`tnext`) to it at every iteration -- mirroring the existing `t1`-clamp pattern in `circuit_diffeq.py`'s loop. This means `assembly.py`'s `tau >= dt` guard becomes a pure safety net that should essentially never fire, rather than something adaptive stepping can trip mid-run. `ConstantStepSize` behavior is completely unchanged: an explicit `dt0 > tau` still raises.

Gradient through the clamp value itself is intentionally stopped (same convention diffrax's own `PIDController` uses for its step-size-control inputs) -- the delayed values actually read from the history buffer remain fully differentiable.

Tests: replaced the old "adaptive+delay raises" test with three new ones -- `PIDController` output matches the analytic time-shifted signal, matches a `ConstantStepSize` run on the same circuit (validating `jnp.interp` accuracy against the now-irregularly-spaced history buffer), and a gradient-vs-finite-difference check under `PIDController`. One thing worth noting: the FD check needed a different evaluation point than the constant-step version -- deep in the settled/saturated region the true derivative is tiny (~1e-7) and gets swamped by mesh-selection noise that differs slightly between the independently-adapted `+h`/`-h` solves (not a gradient bug, just an ill-conditioned comparison for adaptive stepping); moving to the pulse's rising edge, where the signal is large, resolved it cleanly. Full suite (223 tests) passes with zero regressions.

No PR yet -- happy to open one when ready.

**Comment by cdaunt** [comment]
Featured in #49
